File: mps-m/Mpal.Model/Operand.cs
//    MeaProcess - Meaurement and Automation framework.
//    Copyright (C) 2015  Laurentiu-Gheorghe Crisan
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;

using System.Text;

namespace Mpal.Model
{
    public class Operand
    {
        public enum Type
        {
            NoneType = 0,
            Direct,
            Immediate,
            Reference,
            Temporary,
            TemporaryRef,
        }

        private uint     _offset;
        private ulong    _uid;
        private Type     _type;
        private object   _constant;
        private uint     _size;
        private string   _typeId;
        private bool     _isConstant;
        private bool     _cached;

        public Operand()
        {
        }

        public Operand(Operand op)
        {
            _offset = op._offset;
            _uid = op._uid;
            _type = op._type;
            _constant = op._constant;
            _size = op._size;
            _typeId = op._typeId;
            _isConstant = op.IsConstant;
            _cached = op._cached;
        }

        public Operand(uint offset, ulong uid, Type type, uint size, string typeId)
        {
            _offset = offset;
            _uid = uid;
            _type = type;
            _size = size;
            _typeId = typeId;
        }

        public Operand(object constant, ulong uid, uint size, string typeId)
        {
            _constant = constant;
            _type = Type.Immediate;
            _uid = uid;
            _offset = 0;
            _typeId = typeId;
            _size = size;
        }

        public bool IsConstant
        {
            get{ return _isConstant;}
            set { _isConstant = value; }
        }

        public bool Cached
        {
            get { return _cached; }
            set { _cached = value; }
        }
        public string TypeId
        {
            get { return _typeId; }
            set { _typeId = value; }
        }

        public uint Size
        {
            get { return _size; }
            set { _size = value; }
        }

        public uint Offset
        {
            get { return _offset; }
            set { _offset = value; }
        }

        public ulong Uid
        {
            get { return _uid; }
            set { _uid = value; }
        }

        public Type OpType
        {
            get { return _type; }
            set { _type = value; }
        }

        public object ConstVal
        {
            get{return _constant;}
            set { _constant = value; }
        }
    }
}

File: mps-n/src/mpal-vm/Operand.cpp
/*
 * Copyright (C) 2015 by Laurentiu-Gheorghe Crisan
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * As a special exception, you may use this file as part of a free
 * software library without restriction. Specifically, if other files
 * instantiate templates or use macros or inline functions from this
 * file, or you compile this file and link it with other files to
 * produce an executable, this file does not by itself cause the
 * resulting executable to be covered by the GNU General Public
 * License. This exception does not however invalidate any other
 * reasons why the executable file might be covered by the GNU Library
 * General Public License.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "Operand.h"

namespace mpal{
namespace vm{

Operand::Operand(Pt::uint32_t offset,Type type)
: _offset(offset)
, _size(0)
, _const(0)
, _type(type)
{
}

Operand::Operand(Pt::uint64_t constant)
:  _offset(0)
, _size(0)
, _const(constant)
, _type(Immediate)

{
}

Operand::Operand()
: _offset(0)
, _size(0)
, _const(0)
, _type(NoneType)
{
}

Operand::~Operand(void)
{ 

}
void Operand::setOffset(Pt::uint32_t offset)
{
    _offset = offset;
}

void Operand::setSize( Pt::uint32_t size)
{
    _size = size;
}

Pt::uint32_t Operand::size() const
{
    return _size;
}

Pt::uint8_t* Operand::getOperandPtr(Pt::uint8_t* basePtr) const
{
    switch(_type)
    {

        case Direct:
        case Temporary:
            return (Pt::uint8_t*)(basePtr + _offset);

        case Reference:
        case TemporaryRef:
        {
            size_t* ref = (size_t*)(basePtr + _offset);
            return (Pt::uint8_t*) (*ref);
        }

        case Immediate:
        {
            return (Pt::uint8_t*) & _const;
        }
        default:
        throw std::logic_error("Wrong address");
    }

    throw std::logic_error("Wrong address");
    return 0;
}

Pt::uint8_t* Operand::calcAddress(Pt::uint8_t* basePtr, bool address) const
{
    switch(_type)
    {
        case Direct:
        case Temporary:
            return basePtr + _offset;

        case Reference:
        case TemporaryRef:
        {
            size_t* ref = (size_t*)(basePtr + _offset);
                
            if(address)
                return (Pt::uint8_t*) ref;
            else
                return (Pt::uint8_t*) (*ref);
        }

        case Immediate:
        {
            return (Pt::uint8_t*) & _const;
        }
        default:
            throw std::logic_error("Wrong address");
            return 0;
    }

    throw std::logic_error("Wrong address");
    return 0;
}

}}

